bestla/bestla/bestla_parallel.h

```diff
@@ -581,84 +834,131 @@ class SchedulerKBlockS : public SchedulerBase<_GemmCore_T> {
   int mKBlock{0};
 };
 
-}  // namespace gemm
-using thread_func = std::function<void(int tid)>;
-
-class IThreading {
- public:
-  explicit IThreading(int nthreads) : mThreadNum(nthreads) {}
-  virtual void parallel_for(const thread_func& func) const = 0;
-  virtual inline void sync() const { assert(0); };
-  virtual int num_threads() const { return mThreadNum; };
-  virtual void set_threads(int nthreads) = 0;
-
- protected:
-  int mThreadNum;
-};
-#if BTLA_OPENMP
-class OMPThreading : public IThreading {
+template <class Scheduler>
+class SchedulerDispatcher {
  public:
-  explicit OMPThreading(int nthreads) : IThreading(nthreads) { omp_set_num_threads(nthreads); }
-  void parallel_for(const thread_func& func) const override {
-    if (mThreadNum > 1) {
-#pragma omp parallel
-      {
-        int tidx = omp_get_thread_num();
-        func(tidx);
-      }
+  using ThreadProblem = ThreadProblemBase;
+  SchedulerDispatcher() = default;
+  ~SchedulerDispatcher() {
+    std::pair<float, float> PEtime = th_->get_PEtime();
+    if (needDispach && int(PEtime.first) > 0 && int(PEtime.second) > 0)
+      cr->adjustPE(Scheduler::gemm_ISA(), PEtime.second / PEtime.first);
+    delete Scheduler_P;
```

Comment:
use std::shared_ptr instead

Reply:
as it's a high-freq call, use stack memory instead of heap memory

Reply:
done